qualtran/bloqs/cswap_lth_reg.py

```diff
@@ -0,0 +1,122 @@
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from typing import Tuple
+
+import attr
+import cirq
+from cirq._compat import cached_property
+
+from qualtran import bloq_example, BloqDocSpec, Register, SelectionRegister
+from qualtran._infra.gate_with_registers import total_bits
+from qualtran.bloqs.basic_gates import CSwap
+from qualtran.bloqs.unary_iteration_bloq import UnaryIterationGate
+
+
+@attr.frozen
+class ApplyCSwapToLthReg(UnaryIterationGate):
+    r"""Swaps the $l$-th register into an ancilla using unary iteration.
+
+    Applies the unitary which peforms
+    $$
+        U |l\rangle|\psi_0\rangle\cdots|\psi_l\rangle\cdots|\psi_n\rangle|\mathrm{junk}\rangle
+        \rightarrow
+        |l\rangle|\psi_0\rangle\cdots|\mathrm{junk}\rangle\cdots|\psi_n\rangle|\psi_l\rangle
+    $$
+    through a combination of unary iteration and CSwaps.
+
+    The cost should be $L n_b + L - 2 + n_c$, where $L$ is the
+    iteration length, $n_b$ is the bitsize of
+    the registers to swap, and $n_c$ is the number of controls.
+
+    Args:
+        selection_regs: Indexing `select` signature of type Tuple[`SelectionRegisters`, ...].
+            It also contains information about the iteration length of each selection register.
+        bitsize: The size of the registers we want to swap.
+        nth_gate: A function mapping the composite selection index to a single-qubit gate.
+
+    Registers:
+        control_registers: Control registers
+        selection_regs: Indexing `select` signature of type Tuple[`SelectionRegisters`, ...].
+            It also contains information about the iteration length of each selection register.
+        target_registers: Target registers to swap. We swap FROM registers
+            labelled x`i`, where i is an integer and TO a single register called y
+
+    References:
+        [Fault-Tolerant Quantum Simulations of Chemistry in First Quantization](
+            https://arxiv.org/abs/2105.12767) page 20 paragraph 2.
+    """
+    selection_regs: Tuple[SelectionRegister, ...] = attr.field(
+        converter=lambda v: (v,) if isinstance(v, SelectionRegister) else tuple(v)
+    )
+    bitsize: int
+    control_regs: Tuple[Register, ...] = attr.field(
+        converter=lambda v: (v,) if isinstance(v, Register) else tuple(v),
+        default=(Register('ctrl', 0),),
+    )
+
+    @cached_property
+    def control_registers(self) -> Tuple[Register, ...]:
+        return self.control_regs
+
+    @cached_property
+    def selection_registers(self) -> Tuple[SelectionRegister, ...]:
+        return self.selection_regs
+
+    @cached_property
+    def target_registers(self) -> Tuple[Register, ...]:
+        iteration_length = self.selection_registers[0].iteration_length
+        regs = [Register(f'x{i}', bitsize=self.bitsize) for i in range(iteration_length)]
+        regs += [Register('y', bitsize=self.bitsize)]
+        return tuple(regs)
```

Comment:
This assumes there's a single selection register and also does not use the fact that we can support multi-dimensional target registers. Support the more generic case using
```suggestion
        target_shape = tuple(sreg.iteration_length for sreg in self.selection_registers)
        return Signature([Register('targets', bitsize=self.target_bitsize, shape=target_shape), Register('output', bitsize=self.target_bitsize)])
```

Reply:
Done % some changes (shouldn't return a Signature)